source/ada/lsp-ada_documents.adb

```diff
@@ -443,4 +424,20 @@ package body LSP.Ada_Documents is
       end;
    end Get_Completions_At;
 
+   ----------
+   -- Unit --
+   ----------
+
+   function Unit (Self : Document) return Libadalang.Analysis.Analysis_Unit is
+      File : constant LSP.Types.LSP_String :=
+        Self.Context.URI_To_File (Self.URI);
+      Unit : Libadalang.Analysis.Analysis_Unit;
+   begin
+      Unit := Self.LAL.Get_From_Buffer
```

Comment:
Why do we need Self.Buffer? LibAdaLang keeps text by it-self if we have already parsed it. And we parse file on DidOpen request. We can fetch Analysis_Unit at any moment using Self.LAL.Get_From_File.

Reply:
I don't understand this comment: we need to store the buffer in response to textdocument/didChange - we can't Get_From_File for buffers that contain unsaved changes.

Reply:
I mean, we don't need the buffer.

When client opens/change the document, it send us document text and we pass it to LAL providing file name as unique document identifier. At any request we can get Analysis_Unit from LAL. To do this we just call `Get_From_File (File_Name, Reparse => False)`.

> we can't Get_From_File for buffers that contain unsaved changes.

Yes, we can! We have provided the change to LAL at `didChange` processing, so we just call Get_From_File and LAL won't read the file, but return already parsed unit.

All we need to implement this ticket, just drop Unit field from Document record.